### ISSUE TITLE
[BUG] Middleware calls

### ISSUE THREAD

**Issue by ndureiss** [7PH/deadlock.js#42]
Hi @7PH,

Using Deadlock.js, it seems that two points suggest a bug.

First of all, it seems you can define `APIDirectory.middleware` instead of an `APIEndpoint.middlewares` without any compilation error. Can be confusing ...

*Example :*
```ts
// APIDirectory
middleware: [ // APIDirectory.middleware
  async (req: Request, res: Response): Promise<void> => {
    ...some code ...
  },
],
routes: { 
  "/foo": {
    // APIDirectory
    routes: {
      "/bar": {
        // APIEndpoint
        method: "get", 
        middleware: [mwFunc()],  // APIDirectory.middleware
        handler: fooBar(),
      },
    },
  },
}
```

Then, it seems that only defined `APIDirectory.middleware` are handled and executed, `APIEndpoint.middlewares` seems never called...

Thx for investigating :rocket:

**Comment by 7PH** [comment]
> First of all, it seems you can define APIDirectory.middleware instead of an APIEndpoint.middlewares without any compilation error. Can be confusing ...

That's normal since a route (an object within the `routes` field of an APIDirectory) can either be an end point or a directory. TypeScript can't infer what each type is so it will treat each route as `Union<APIDirectory | APIEndPoint | APIEndPointHandler>`.

> Then, it seems that only defined APIDirectory.middleware are handled and executed, APIEndpoint.middlewares seems never called...

That's also normal because it's APIEndpoint.middlewar**e** not APIEndpoint.middlewar**es**
As far as I understand:
- The naming should be the same for end points and directories (`middleware`) so there is no confusion between the two. I'll fix this.
- An additional feature that would be interesting is running a check at runtime that no useless property is defined and throw an error if that's the case.

If you want type-checking in your IDE you can use this:
```typescript
// APIDirectory
routes: { 
  "/foo": {
    middleware: [
      async (req: Request, res: Response): Promise<void> => { },
    ],
    routes: {
      "/bar": {
        method: "get", 
        middlewares: [mwFunc()],
        handler: fooBar(),
      } as APIEndPoint,
    },
  } as APIDirectory,
}
```

Or even
```typescript
// APIDirectory
routes: { 
  "/foo": <APIDirectory> {
    middleware: [
      async (req: Request, res: Response): Promise<void> => { },
    ],
    routes: {
      "/bar": <APIEndPoint> {
        method: "get", 
        middlewares: [mwFunc()],
        handler: fooBar(),
      },
    },
  },
}
```

**Comment by ndureiss** [comment]
> That's normal since a route (an object within the `routes` field of an APIDirectory) can either be an end point or a directory. TypeScript can't infer what each type is so it will treat each route as `Union<APIDirectory | APIEndPoint | APIEndPointHandler>`.

I didn't have the `Union`, for me it's was a simple OR :smile: 

> > Then, it seems that only defined APIDirectory.middleware are handled and executed, APIEndpoint.middlewares seems never called...
> 
> That's also normal because it's APIEndpoint.middlewar**e** not APIEndpoint.middlewar**es**

I understand my problem! In fact my middleware is called *after* the endpoint handler while I expect *before*, is there any way to run the middleware before the endpoint handler ?

> As far as I understand:
> 
> * The naming should be the same for end points and directories (`middleware`) so there is no confusion between the two. I'll fix this.

Could be useful, thanks :+1: 

> * An additional feature that would be interesting is running a check at runtime that no useless property is defined and throw an error if that's the case.
> 
> If you want type-checking in your IDE you can use this:
> 
> ```ts
> routes: { 
>   "/foo": {
>     middleware: [
>       async (req: Request, res: Response): Promise<void> => { },
>     ],
>     routes: {
>       "/bar": {
>         method: "get", 
>         middlewares: [mwFunc()],
>         handler: fooBar(),
>       } as APIEndPoint,
>     },
>   } as APIDirectory,
> }
> ```

Could be a good mitigation. Thanks !

**Comment by 7PH** [comment]
> I understand my problem! In fact my middleware is called after the endpoint handler while I expect before, is there any way to run the middleware before the endpoint handler ?

Ok I get it, looks like there is a bug.
This snippet:
```typescript
import * as express from "express";

import {APIDescription} from "../src";
import {DeadLock} from "../src/deadlock";

const HOST: string = 'localhost';
const PORT: number = 48654;
const PATH: string = '/api/test';

const api: APIDescription = {

    verbose: true,

    hostname: HOST,

    port: PORT,

    basePath: PATH,

    routes: {

        '/dir': {
            middleware: [
                async (req: express.Request, res: express.Response): Promise<void> => {
                    console.log('directory middleware 1');
                },
                async (req: express.Request, res: express.Response): Promise<void> => {
                    console.log('directory middleware 2');
                }
            ],
            routes: {
                '/route': {
                    middlewares: [
                        async (req: express.Request, res: express.Response): Promise<void> => {
                            console.log('end point middleware 1');
                        },
                        async (req: express.Request, res: express.Response): Promise<void> => {
                            console.log('end point middleware 2');
                        },
                    ],
                    handler: async () => {
                        console.log('handler');
                        return "returned value";
                    }
                }
            }
        }
    }
};

(async () => {

    await DeadLock.startApp(api);
})();
```

Outputs, when navigating to http://localhost:48654/api/test/dir/route:
```text
directory middleware 1
directory middleware 2
handler
end point middleware 1
end point middleware 2
```

Instead of
```text
directory middleware 1
directory middleware 2
end point middleware 1
end point middleware 2
handler
```

Middlewares should be run BEFORE the handler of course. I'll look this up

**Comment by 7PH** [comment]
Fixed in `1.5.0`, the snippet above now shows:
```text
directory middleware 1
directory middleware 2
end point middleware 1
end point middleware 2
handler
```

Also `middleware` has been renamed to `middlewares`, that's a breaking change so beware

**Comment by ndureiss** [comment]
Thx for the fix ! 💪
